### PR TITLE
feat: CPU metric collection

### PR DESCRIPTION
**Pull request by GustavBaumgart**
## Description

Metric collector now uses a thread to collect information on CPU usage. The data points are (currently) taken every 10 seconds, but this can be adjusted later.

If this layout appears to work well, a similar structure will be used to collect GPU usage metrics, although it will involve more details.

## Type of Change

- [ ] Bug Fix
- [x] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [ ] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [ ] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: lib/python/flame/monitor/metric_collector.py [hunk 1]
```diff
@@ -16,13 +16,34 @@
...
+        
+    
+    def cpu_stats(self, freq=10):
+        # as of now, this import is not in the requirements for flame
+        import psutil
+        pid = os.getpid()
+        proc = psutil.Process(pid)
+        
+        while True:
+            self.points['cpu_utilization'].append((time.time(), proc.cpu_percent()))
```
Comment (jaemin-shin) [inline lib/python/flame/monitor/metric_collector.py:42]: maybe we need better interpretable name than points? how about log or stat_log?

File: lib/python/flame/monitor/metric_collector.py [hunk 2]
```diff
@@ -16,13 +16,34 @@
...
         """Initialize Metric Collector."""
         self.state_dict = dict()
+        self.points = defaultdict(list)
+        cpu_thread = threading.Thread(target=self.cpu_stats)
+        cpu_thread.start()
+        
+    
+    def cpu_stats(self, freq=10):
+        # as of now, this import is not in the requirements for flame
+        import psutil
```
Comment (jaemin-shin) [inline lib/python/flame/monitor/metric_collector.py:37]: why is it imported here?

File: lib/python/flame/monitor/metric_collector.py [hunk 3]
```diff
@@ -16,13 +16,34 @@
...
+        pid = os.getpid()
+        proc = psutil.Process(pid)
+        
+        while True:
+            self.points['cpu_utilization'].append((time.time(), proc.cpu_percent()))
+            
+            self.points['cpu_memory'].append((time.time(), proc.memory_info().rss))
```
Comment (jaemin-shin) [inline lib/python/flame/monitor/metric_collector.py:44]: do you know how much overhead are involved in measuring cpu_utilization and cpu_memory in given method?

File: lib/python/flame/monitor/metric_collector.py [hunk 4]
```diff
@@ -38,9 +59,23 @@ def accumulate(self, mtype, alias, value):
         self.state_dict[key] = value + self.state_dict.get(key, 0)
         logger.debug(f"Accumulating metric state_dict[{key}] = {self.state_dict[key]}")
     
+    def process_points(self):
```
Comment (jaemin-shin) [inline lib/python/flame/monitor/metric_collector.py:62]: I think you could also use more interpretable name here too, "process" itself could be abstract. How about something like "save_points_statistics"?

File: lib/python/flame/monitor/metric_collector.py (hunk 1 above)
Comment (jaemin-shin) [reply lib/python/flame/monitor/metric_collector.py:42]: also, just out of curiosity, is this supported on other OS than ubuntu?

**Review by jaemin-shin — COMMENTED** [review]
Overall LGTM, left some minor suggestions and questions.

File: lib/python/flame/monitor/metric_collector.py (hunk 2 above)
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:37]: I thought it would be better since psutil is not currently included in the requirements, so it would be easier to toggle it on and off if it is there and not installed. Still haven't added a functionality for how to switch metric collection on and off.

File: lib/python/flame/monitor/metric_collector.py (hunk 1 above)
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:42]: Hmm I agree with the idea that there's a better name, but I'm not sure stat_log suggests x,y coordinates. As of now, it does look like other operating systems would be able to run that as well.

File: lib/python/flame/monitor/metric_collector.py (hunk 3 above)
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:44]: No idea.

File: lib/python/flame/monitor/metric_collector.py (hunk 1 above)
Comment (jaemin-shin) [reply lib/python/flame/monitor/metric_collector.py:42]: OS-diagnostic is good. It's still a minor thing, but I am still wondering why that "point" variable have a meaning of x, y coordinate in its name? It could be that if you consider a graph, but more directly, I thought it's just a tuple of time and the measurement value of metric that we want to collect.

File: lib/python/flame/monitor/metric_collector.py (hunk 2 above)
Comment (jaemin-shin) [reply lib/python/flame/monitor/metric_collector.py:37]: I see. As we do not have separate set of pip requirements to install for new users, I thought it could be much simple to just ask all users to do pip install psutils and import it along with other packages at the beginning of this file. I would like to hear what others think too.

File: lib/python/flame/monitor/metric_collector.py (hunk 3 above)
Comment (jaemin-shin) [reply lib/python/flame/monitor/metric_collector.py:44]: It's good that psutils are efficient way to measure it, but I think we should have some sense and be able to recommend of which frequency would be appropriate to measure these accurately without overhead from this itself.
